### DISCUSSION TITLE
Pattern variables

### DISCUSSION THREAD

**Discussion by kyoyama-kazusa** [dotnet/csharplang#4776] [Language Ideas]
Hello.

Can we extract a pattern into a variable to store, and then use it multiple times? Such as:

```csharp
expr equalOrNotEquals = not ((int)SyntaxKind.EqualsExpression or (int)SyntaxKind.NotEqualsExpression);

if (
    node is BinaryExpression
    {
        RawKind: (int)SyntaxKind.LogicalAndExpression,
        Left: { RawKind: not equalOrNotEquals },
        Right: { RawKind: equalOrNotEquals }
    }
)
{
    // ...
}
```

**Comment by 333fred** [comment]
~~Replace `expr` with `const int` and yes, you can.~~

When you answer things later at night, sometimes you misread.

**Reply by kyoyama-kazusa** [reply]
Hmmm... What about other patterns, such as property pattern? A property pattern isn't a constant, I think.

**Reply by 333fred** [reply]
Hmm. I entirely misread your question, ignore my answer.

**Comment by alrz** [comment]
Does https://github.com/dotnet/csharplang/issues/4131 address your question?

**Reply by kyoyama-kazusa** [reply]
That proposal does make sense to me. Maybe that proposal has already covered the idea here. But... as a beginner for learning the pattern matching, using method invocation as a pattern is a little bit weird, especially combined with other pattern types (only in my opinion).

Anyway, thanks for your reply.

**Comment by WhiteBlackGoose** [comment]
I would have syntax like
`pattern {TYPE TO BE EXAMINED} {NAME OF PATTERN} {PATTERN}`
For example,
```cs
public static class MyClass
{
    pattern int Positive > 0;
    pattern int Even % 2 == 0;
    pattern Complex Real { RealPart: > 0 };
}
```
example similar to what the OP posted:
```cs
if (c is Real re) ...
if (a is Even) ...
if (student is { Attitude: Positive }) ...
```
(the patterns will be colored differently from types by IDEs to make it clear)
@333fred what do you think?

**Reply by WhiteBlackGoose** [reply]
Here's how we could handle cases with extracting variables:
```cs
pattern Person ValidPerson(name) { Age: > 0, Name: { } name }
```
Use case:
```cs
if (person is ValidPerson(var name)) ...
```

**Reply by 333fred** [reply]
I think @alrz's proposal or Neal's original one are the next steps to go in.
